curp/src/client.rs

```diff
@@ -1,36 +1,111 @@
         ReadState as PbReadState, SyncResult, WaitSyncedRequest,
     },
     LogIndex, ServerId,
 };
 
 /// Protocol client
+#[derive(derive_builder::Builder)]
+#[builder(build_fn(skip), name = "Builder")]
 pub struct Client<C: Command> {
     /// local server id. Only use in an inner client.
+    #[builder(field(type = "Option<ServerId>"), setter(custom))]
     local_server_id: Option<ServerId>,
     /// Current leader and term
+    #[builder(setter(skip))]
     state: RwLock<State>,
-    /// All servers's `Connect`
-    connects: HashMap<ServerId, Arc<dyn ConnectApi>>,
     /// Curp client timeout settings
     timeout: ClientTimeout,
     /// To keep Command type
+    #[builder(setter(skip))]
     phantom: PhantomData<C>,
 }
 
+impl<C: Command> Builder<C> {
+    /// Set local server id.
+    #[inline]
+    pub fn local_server_id(&mut self, value: ServerId) -> &mut Self {
+        self.local_server_id = Some(value);
+        self
+    }
+
+    /// Build client from all members
+    /// # Errors
+    /// Return error when meet rpc error or missing some arguments
+    #[inline]
+    pub async fn build_from_all_members(
+        &self,
+        all_members: HashMap<ServerId, String>,
+    ) -> Result<Client<C>, ClientBuildError> {
+        let Some(timeout) = self.timeout else {
+            return Err(ClientBuildError::invalid_aurguments("timeout is required"));
+        };
+        let connects = rpc::connect(all_members).await;
+        let client = Client::<C> {
+            local_server_id: self.local_server_id.clone(),
+            state: RwLock::new(State::new(None, 0, connects)),
+            timeout,
+            phantom: PhantomData,
+        };
+        Ok(client)
+    }
+
+    /// Build client from addresses, this method will fetch all members from servers
+    /// # Errors
+    /// Return error when meet rpc error or missing some arguments
+    #[inline]
+    pub async fn build_from_addrs(
+        &self,
+        addrs: Vec<String>,
+    ) -> Result<Client<C>, ClientBuildError> {
+        let Some(timeout) = self.timeout else {
+            return Err(ClientBuildError::invalid_aurguments("timeout is required"));
+        };
+        let futs = addrs.into_iter().map(|mut addr| {
+            if !addr.starts_with("http://") {
+                addr.insert_str(0, "http://");
+            }
+            let timeout = *timeout.propose_timeout();
+            let fut = async move {
+                let mut protocol_client = ProtocolClient::connect(addr).await?;
+                let mut req = tonic::Request::new(FetchClusterRequest::new());
+                req.set_timeout(timeout);
+                let fetch_cluster_res = protocol_client.fetch_cluster(req).await?.into_inner();
+                Ok::<FetchClusterResponse, ClientBuildError>(fetch_cluster_res)
+            };
+            Box::pin(fut)
+        });
+
+        let res = select_all(futs).await.0?;
```

Comment:
As the operation in `fut` may fail, we should consider select the first successful future that returned. If none of them success, return an error instead.

Reply:
Fixed